service/history/api/consistency_checker.go

```diff
@@ -239,45 +240,44 @@ func (c *WorkflowConsistencyCheckerImpl) getCurrentWorkflowContext(
 	namespaceID string,
 	workflowID string,
 ) (WorkflowContext, error) {
-	for attempt := 1; attempt <= conditionalRetryCount; attempt++ {
-		runID, err := c.getCurrentRunID(
-			ctx,
-			shardOwnershipAsserted,
-			namespaceID,
-			workflowID,
-		)
-		if err != nil {
-			return nil, err
-		}
-		wfContext, err := c.getWorkflowContextValidatedByCheck(
-			ctx,
-			shardOwnershipAsserted,
-			consistencyPredicate,
-			definition.NewWorkflowKey(namespaceID, workflowID, runID),
-		)
-		if err != nil {
-			return nil, err
-		}
-		if wfContext.GetMutableState().IsWorkflowExecutionRunning() {
-			return wfContext, nil
-		}
+	runID, err := c.getCurrentRunID(
+		ctx,
+		shardOwnershipAsserted,
+		namespaceID,
+		workflowID,
+	)
+	if err != nil {
+		return nil, err
+	}
+	wfContext, err := c.getWorkflowContextValidatedByCheck(
+		ctx,
+		shardOwnershipAsserted,
+		consistencyPredicate,
+		definition.NewWorkflowKey(namespaceID, workflowID, runID),
+	)
+	if err != nil {
+		return nil, err
+	}
+	if wfContext.GetMutableState().IsWorkflowExecutionRunning() {
+		return wfContext, nil
+	}
 
-		currentRunID, err := c.getCurrentRunID(
-			ctx,
-			shardOwnershipAsserted,
-			namespaceID,
-			workflowID,
-		)
-		if err != nil {
-			wfContext.GetReleaseFn()(err)
-			return nil, err
-		}
-		if currentRunID == wfContext.GetRunID() {
-			return wfContext, nil
-		}
-		wfContext.GetReleaseFn()(nil)
+	currentRunID, err := c.getCurrentRunID(
+		ctx,
+		shardOwnershipAsserted,
+		namespaceID,
+		workflowID,
+	)
```

Comment:
why we need this get again? It is done at line 243 already

Reply:
If workflow is closed, we can't know if there's a newer run created after getting the runID on L243. So need to get runID again to validate it's still the latest run. And since workflow is locked before the second check, there's won't be a newer run if the second check is successful.